File: backend/src/service/agent-tools/tags.tools.ts
import { tool } from "langchain";
import { z } from "zod";
import { getTagsList, getArticlesByTag } from '@/service/tags.service';

// 工具1：获取标签列表
export const getTagsListTool = tool(
    async ({ page, size }) => {
        const tags = await getTagsList(page, size);
        return JSON.stringify({
            tags: tags.map(tag => ({
                name: tag.name,
                articleCount: tag.articleCount
            })),
            note: "请根据上述标签列表，用自然语言向用户介绍各个标签及其文章数量，不要直接显示 JSON。"
        });
    },
    {
        name: "get_tags_list",
        description: "获取文章标签列表及每个标签下的文章数量",
        schema: z.object({
            page: z.number().default(1).describe("页码"),
            size: z.number().default(10).describe("每页数量"),
        }),
    }
);

// 工具2：根据标签获取文章
export const getArticlesByTagTool = tool(
    async ({ tagName, page, size }) => {
        const articles = await getArticlesByTag(tagName, page, size);
        return JSON.stringify({
            tag: tagName,
            articles: articles.map(article => ({
                title: article.title,
                content: article.content?.substring(0, 100) + '...'
            })),
            note: "请根据上述文章列表，用自然语言向用户推荐相关文章，不要直接显示 JSON。"
        });
    },
    {
        name: "get_articles_by_tag",
        description: "根据标签名称获取相关文章列表",
        schema: z.object({
            tagName: z.string().describe("标签名称"),
            page: z.number().default(1).describe("页码"),
            size: z.number().default(10).describe("每页数量"),
        }),
    }
);

// 导出标签相关工具数组
export const tagsTools = [getTagsListTool, getArticlesByTagTool];

File: backend/ai-service/tools/tags_tools.py
import json

import httpx
from langchain_core.tools import tool

import config
from tools.java_api import extract_list

JAVA_URL = config.JAVA_BACKEND_URL


@tool
async def get_tags_list() -> str:
    """获取博客所有标签列表。当用户问"有哪些标签"、"文章都有什么分类标签"时使用此工具。"""
    async with httpx.AsyncClient(timeout=config.HTTP_TIMEOUT) as client:
        resp = await client.get(f"{JAVA_URL}/api/tags")
        return json.dumps({
            "tags": extract_list(resp.json()),
        }, ensure_ascii=False)


@tool
async def get_articles_by_tag(name: str) -> str:
    """根据标签名称获取相关文章列表。当用户问"XX标签下有哪些文章"、"带某标签的文章"时使用此工具。"""
    async with httpx.AsyncClient(timeout=config.HTTP_TIMEOUT) as client:
        resp = await client.get(f"{JAVA_URL}/api/tags/{name}")
        clean_data = [
            {"title": a.get("title"), "content": a.get("content")}
            for a in extract_list(resp.json())
        ]
        return json.dumps({
            "tag": name,
            "articles": clean_data,
        }, ensure_ascii=False)


tags_tools = [get_tags_list, get_articles_by_tag]
